File: python/151-200/170. Two Sum III - Data structure design.py
class TwoSum(object):

    def __init__(self):
        """
        initialize your data structure here
        """
        self.temp = {}
    def add(self, number):
        """
        Add the number to an internal data structure.
        :rtype: nothing
        """
        if number not in self.temp:
            self.temp[number] = 1
        else:
            self.temp[number] += 1
    def find(self, value):
        """
        Find if there exists any pair of numbers which sum is equal to the value.
        :type value: int
        :rtype: bool
        """
        for v in self.temp:
            if value - v in self.temp and (value - v != v or self.temp[v] > 1):
                return True
        return False

File: cpp/151-200/170. Two Sum III - Data structure design.cc

class TwoSum {
private:
    unordered_map<int, int> dict;
public:
    /** Initialize your data structure here. */
    TwoSum() {
    }

    /** Add the number to an internal data structure.. */
    void add(int number) {
        dict[number]++;
    }

    /** Find if there exists any pair of numbers which sum is equal to the value. */
    bool find(int value) {
        for (auto it : dict) {
        	int tmp = value - it.first;
        	if(tmp != it.first & dict.count(tmp)){
        		return true;
        	}
        	if(tmp == it.first & it.second >= 2)
        		return true;
        }
        return false;
    }
};

/**
 * Your TwoSum object will be instantiated and called as such:
 * TwoSum obj = new TwoSum();
 * obj.add(number);
 * bool param_2 = obj.find(value);
 */